### DISCUSSION TITLE
Be able to use a list to define the pins for ButtonBoard

### DISCUSSION THREAD

**Discussion by Kayakbabe** [gpiozero/gpiozero#1053] [General]
Short description of change: be able to define ButtonBoard pins with a list
Where would this change be made: ButtonBoard

Purpose of change:
Allow defining a ButtonBoard from a configuration list of values instead of having to hardcode them.

Please provide example usage of gpiozero with your suggested change: 

```python

from gpiozero import ButtonBoard

buttons_config = {
6: "next",
5: "prev",
19:"up",
26:"down",
0:"play",
16:"home",
21:"hiddenScreen",
17:"reboot"
}

#bb = ButtonBoard(19,26,4,hold_time=3,hold_repeat = False)

bb = ButtonBoard(buttons_config.keys(),hold_time=3,hold_repeat = False)

# or 
myPins = buttons_config.keys()
bb = ButtonBoard(myPins,hold_time=3,hold_repeat = False)


```

**Comment by lurch** [comment]
I think `bb = ButtonBoard(*buttons_config.keys(),hold_time=3,hold_repeat = False)` should do what you want? :slightly_smiling_face: 

Alternatively, you could 'invert' your buttons_config dictionary, and do something like:
```python3
from gpiozero import ButtonBoard

buttons_config = {
   "next": 6,
   "prev": 5,
   "up": 19,
   "down": 26,
   "play": 0,
   "home": 16,
   "hiddenScreen": 21,
   "reboot": 17
}


bb = ButtonBoard(hold_time=3, hold_repeat=False, **buttons_config)
```

See https://docs.python.org/3/tutorial/controlflow.html#unpacking-argument-lists

**Comment by bennuttall** [comment]
Yes, what @lurch said.

If you have a list on pin numbers, unpack it with `*` like so: `ButtonBoard(*button_pins)` which is the same as e.g. `ButtonBoard(2, 3, 4, 5)`

If you have a dict mapping names to pins (the opposite of what you have), even better because you can do `ButtonBoard(**button_pins)` which is the same as `ButtonBoard(next=6, prev=5, up=19, down=26)` and then you can access them by name like so:

```
bb.next.when_pressed = do_next
bb.prev.when_pressed = do_prev
```

This doesn't seem to be covered in the docs but the `LEDBoard` and `StatusBoard` examples are similar: https://gpiozero.readthedocs.io/en/stable/recipes_advanced.html#who-s-home-indicator

**Comment by Kayakbabe** [comment]
Mapping names to pins is nice for one function pins. The **kwargs works but doesn't exactly do what I need. It forces me to define each and every single button of a defined number of buttons I might have in the script. This is good if I were putting out a multi button device with a fixed number of buttons in the world that just did one thing only and wouldn't be configurable. I could name the pins button1, button2 etc.

I added print statements to see what values I get out.

```python
//works but not exactly what i want
from gpiozero import ButtonBoard

buttons_config = {
   "next": 19,
   "prev": 26,
}

bb = ButtonBoard(hold_time=3, hold_repeat=False, **buttons_config)

def do_pressed(this):
    print("PRESSED: "  + " value " + str(bb.value))
    print(this)
    print(this.value)
    print(this.pin)
    print(this.pin.number)
    print(' ')

def do_held(this):
    print("HELD: " + str(bb.held_time) + " value " + str(bb.value))
    print(this)
    print(this.value)
    print(this.pin)
    print(this.pin.number)
    print(' ')

def do_released(this):
    print("RELEASED"  + " value bb:" + str(bb.value))
    print(this)
    print(this.value)
    print(this.pin)
    print(this.pin.number)
    print(' ')

bb.next.when_pressed = do_pressed
bb.next.when_held = do_held
bb.next.when_released = do_released

bb.prev.when_pressed = do_pressed
bb.prev.when_held = do_held
bb.prev.when_released = do_released
```
But if I don't know how many buttons there are going to be, could be 1 or many, I think Lurch's first example using *args is more what I need. I want to be able to pass both the gpio pins and a list of things the buttons to do in a config.js without having to change the script I write if the number of buttons change.

What I want to do is the following but I found that ButtonBoard doesn't return the pin(s) or pin.number(s) when used this way. it throws
_AttributeError: no such attribute_ errors instead. Is there some way to get this info from ButtonBoard?

```python
// kinda works but I need pin.number and it doesn't exist
from gpiozero import ButtonBoard

buttons_config = {
    "19":["next","admin"],
    "26":["prev","softReboot"],
    "17":["screenon","screenoff"]
}

bb = ButtonBoard(*buttons_config,hold_time=3,hold_repeat = False)

def do_pressed(this):
    print("PRESSED "  + " value " + str(bb.value))
    print(this)
    print(this.value)
    print(this.pin)
    print(this.pin.number)
    print("I will get the first value for the key and do something but I need the pin.number to get the value")
    print(' ')

def do_held(this):
    print("HELD: " + str(bb.held_time) + " value " + str(bb.value))
    print(this)
    print(this.value)
    print(this.pin)
    print(this.pin.number)
    print("I will get the second value for the key and do something but I need the pin.number to get the value")
    print(' ')

def do_repleased(this):
    print("HELD: " + str(bb.held_time) + " value " + str(bb.value))
    print(this)
    print(this.value)
    print(this.pin)
    print(this.pin.number)
    print("I will get the third value for the key and do something but I need the pin.number to get the value")
    print(' ')

bb.when_pressed = do_pressed
bb.when_held = do_held
bb.when_released = do_held
```

**Reply by bennuttall** [reply]
When setting events on the whole buttonboard, it doesn't fire every time a button is pressed, only when it goes from 0 pressed to 1 pressed. Instead you might be better just using a loop. Also you can look up the name from the number if you store it that way around:

```python
buttons_config = {
   19: "next",
    26: "prev",
}

def pressed(btn):
    print(buttons_config[btn.pin.number], "button was pressed")

buttons = []
for pin in buttons_config:
    btn = Button(pin)
    btn.when_pressed = pressed
    buttons.append(btn)
```

**Reply by lurch** [reply]
Yeah, IIRC I created ButtonBoard so that you could use the source-value-tools to connect the output of a ButtonBoard to the input of an LEDBoard. But for what you're trying to do (where each button is being treated individually), I think you might be better off with Ben's suggestion of just using lists of Button objects? :shrug: 

Or you could even do this, if you want to be able to more easily modify the buttons after they're created:
```python3
buttons_config = {
   19: "next",
    26: "prev",
}

def pressed(btn):
    print(buttons_config[btn.pin.number], "button was pressed")

buttons_dict = {}
for pin in buttons_config:
    btn = Button(pin)
    btn.when_pressed = pressed
    buttons_dict[pin] = btn
```

**Reply by Kayakbabe** [reply]
I understand now. ButtonBoard was never meant to work like I thought.
Thank you for teaching and helping me. I have a lot to learn about using libraries in python.

Here is what I have ended up with. I'll need to tweak it so held won't trigger pressed. 
but this is proof of concept.

```
from gpiozero import Button
from signal import pause

buttons_config = {
    19:["next","admin"],
    26:["prev","softReboot"],
    17:["screenof","hardReboot"]
}

def do_pressed(btn):
    print("PRESSED: " + str(btn.pin))
    print(buttons_config[btn.pin.number], "button was pressed")
    print("action for " + str(btn.pin.number) + " is")
    thislist = list(buttons_config[btn.pin.number])
    print (thislist[0])
    print(' ')
    
def do_held(btn):
    print("HELD: " + str(btn.pin) )
    print(buttons_config[btn.pin.number], "button was held")
    print("action for " + str(btn.pin.number) + " is")
    thislist = list(buttons_config[btn.pin.number])
    print (thislist[1])
    print(' ')

buttons = []
for pin in buttons_config:
    btn = Button(pin)
    btn.when_pressed = do_pressed
    btn.when_held = do_held
    buttons.append(btn)
```

**Reply by lurch** [reply]
> I'll need to tweak it so held won't trigger pressed.

You'll need to change the code to react on 'released' instead of 'pressed'.
See https://gpiozero.readthedocs.io/en/stable/faq.html#how-do-i-use-button-when-pressed-and-button-when-held-together
